### PR TITLE
Hadoop dist cache should skip GCS upload with direct runner #143

### PR DESCRIPTION
**Pull request by andrewsmartin**

### CODE REVIEW

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala [hunk 1]
```diff
@@ -158,16 +155,35 @@ package object hdfs {
       }
     }
 
-    private[scio] def hadoopDistCacheCopy(src: Path, target: URI, conf: Configuration): Unit = {
-      logger.debug(s"Will copy ${src.toUri}, to $target")
+    private def getTargetFromOptions() = {
+      if (self.options.getRunner() != classOf[DirectPipelineRunner]) {
+        val dfOptions = self.optionsAs[DataflowPipelineOptions]
+        require(self.optionsAs[DataflowPipelineOptions].getStagingLocation != null,
```
Comment (ravwojdyla) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala]: I know this code was here before - but is it needed? From doc:
```
   * <p>At least one of {@link #getTempLocation()} or {@link #getStagingLocation()} must be set. If
   * {@link #getTempLocation()} is not set, then the Dataflow pipeline defaults to using
   * {@link #getStagingLocation()}.
```

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala [hunk 2]
```diff
@@ -158,16 +155,35 @@ package object hdfs {
...
+          "Staging directory not set - use `--stagingLocation`!")
+        (dfOptions.getStagingLocation, gcsOutputStream _)
+      } else {
+        // should targetDir be specified in options?
```
Comment (ravwojdyla) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala]: What do you mean by this comment?

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala (hunk 1 above)
Comment (andrewsmartin) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala]: Oh nice, so we can remove the require and just manually check `tempLocation` and `stagingLocation`?

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala (hunk 2 above)
Comment (andrewsmartin) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala]: I just mean that the "staging location" here is created with `Files.createTempDirectory` using its default directory - just wondering if we should have this configurable via Options and require it be set. But this is just for testing really so it's probably not an issue to just do this.

Comment (ravwojdyla) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala]: It seems like this code to lives in `com.google.cloud.dataflow.sdk.runners.DataflowPipelineRunner.fromOptions`, it handles making sure that staging or temp is set, and if needed sets the missing values. Will this code run after or before `fromOptions` ^ ?

Comment (ravwojdyla) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala]: Actually we shouldn't depend on when this code runs, probably not safe thing to do. So maybe it's just fine to assume that staging has to be set.

**Comment by ravwojdyla** [comment]
👍

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala [hunk 3]
```diff
@@ -28,6 +29,7 @@ import com.google.cloud.dataflow.sdk.coders.AvroCoder
 import com.google.cloud.dataflow.sdk.io.hdfs._
 import com.google.cloud.dataflow.sdk.io.{Read, Write}
 import com.google.cloud.dataflow.sdk.options.DataflowPipelineOptions
+import com.google.cloud.dataflow.sdk.runners.DirectPipelineRunner
```
Comment (nevillelyh) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala]: I don't see this import being used?

Comment (andrewsmartin) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala]: Yeah, was checking directly against `DirectPipelineRunner` before switching to use `ScioUtil.isLocalRunner` and this was left in there. Will remove.

**Pull request merged**